arbor/include/arbor/morph/mprovider.hpp

```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+#include <string>
+#include <unordered_map>
+
+#include <arbor/morph/embed_pwlin1d.hpp>
+#include <arbor/morph/primitives.hpp>
+#include <arbor/morph/label_dict.hpp>
+#include <arbor/util/either.hpp>
+
+namespace arb {
+
+using concrete_embedding = embed_pwlin1d;
+
+struct mprovider {
```

Comment:
As fas as I could tell, there were three approaches that could work, while keeping a const-after-initialization interface:

1. What's here, with mutation at init time.
2. A virtualized `mprovider` interface; `init()` then thingifies everything against a different implementation of this interface which detects circular references, and then pulls out what it needs.
3. Introspection on locset and region expressions.

I thought option 3 was going to be too invasive (it would be a fairly heavy reworking of the locset/region code), and that option 2 complicated the interface only to support this one piece of ultimately behind-the-scenes functionality.

Reply:
> Does this boil down to a workaround of `thingify` inside the constructor, using the partially constructed `provider` for one of its arguments?
> 
> If so, can we have private member functions that perform the `try_lookup` guff and are only called in the constructor? That way the `region()` and `locset()` public members can do a simple const-friendly lookup.

They can't be 'really' const, because they need to be able to update the concrete maps during initialization (short of moving to e.g. option (2) above).